core/network/impl/synchronizer_impl.cpp

```diff
@@ -1687,12 +771,273 @@ namespace kagome::network {
+          return true;
+        });
+      }
+    }
+    applyNextBlock();
+  }
+
+  void SynchronizerImpl::visitAncestry(const BlockInfo &root, const auto &f) {
+    std::deque<BlockInfo> stack{root};
+    while (not stack.empty()) {
+      auto block = stack.back();
+      stack.pop_back();
+      switch (f(known_blocks_.at(block.hash))) {
+        case VisitAncestryResult::CONTINUE: {
+          for (auto [it, end] = ancestry_.equal_range(block); it != end; ++it) {
+            stack.emplace_back(it->second);
+          }
+          break;
+        }
+        case VisitAncestryResult::STOP: {
+          return;
+        }
+        case VisitAncestryResult::IGNORE_SUBTREE: {
+          break;
+        }
+      }
+    }
+  }
+
+  bool SynchronizerImpl::canFetchMore(size_t &fetching) const {
+    return fetching < max_parallel_downloads_;
+  }
+
+  bool SynchronizerImpl::validate(const BlocksRequest &request,
+                                  BlocksResponse &response) {
+    auto &blocks = response.blocks;
+    if (blocks.empty()) {
+      return false;
+    }
+    auto need_header = has(request.fields, BlockAttribute::HEADER);
+    auto need_body = has(request.fields, BlockAttribute::BODY);
+    if (auto *hash = boost::get<BlockHash>(&request.from)) {
+      if (blocks[0].hash != *hash) {
+        return false;
+      }
+    } else {
+      auto number = boost::get<BlockNumber>(request.from);
+      if (blocks[0].header.has_value() and blocks[0].header->number != number) {
+        return false;
+      }
+    }
+    if (request.direction == Direction::ASCENDING) {
+      std::ranges::reverse(blocks);
+    }
+    for (auto &block : blocks) {
+      if (block.header.has_value() != need_header) {
+        return false;
+      }
+      if (not need_body and block.body.has_value()) {
+        return false;
+      }
+      if (need_header) {
+        calculateBlockHash(block.header.value(), *hasher_);
+        if (block.hash != block.header->hash()) {
+          return false;
+        }
+      }
+    }
+    if (need_header) {
+      std::optional<BlockInfo> expected;
+      size_t count = 0;
+      for (auto &block : blocks) {
+        auto &header = block.header.value();
+        ++count;
+        if (expected.has_value() and header.blockInfo() != expected) {
+          return false;
+        }
+        expected = header.parentInfo();
+        if (not expected.has_value() and count != blocks.size()) {
+          return false;
+        }
+      }
     }
+    if (need_body) {
+      size_t count = 0;
+      for (auto &block : blocks) {
+        ++count;
+        if (not block.body.has_value()) {
+          if (count != blocks.size()) {
+            return false;
+          }
+          continue;
+        }
+        if (auto known = entry(known_blocks_, block.hash)) {
+          if (not checkExtrinsicRoot(known->data.header.value(), *block.body)) {
+            return false;
+          }
+        }
+      }
+    }
+    std::ranges::reverse(blocks);
```

Comment:
why reverse is needed here?

Reply:
Can we use `blocks` range, just with reversing adapter if needed?